src/wallet/asyncrpcoperation_shieldcoinbase.cpp

```diff
@@ -43,95 +41,47 @@
 
 using namespace libzcash;
 
-static int find_output(UniValue obj, int n) {
-    UniValue outputMapValue = find_value(obj, "outputmap");
-    if (!outputMapValue.isArray()) {
-        throw JSONRPCError(RPC_WALLET_ERROR, "Missing outputmap for JoinSplit operation");
-    }
-
-    UniValue outputMap = outputMapValue.get_array();
-    assert(outputMap.size() == ZC_NUM_JS_OUTPUTS);
-    for (size_t i = 0; i < outputMap.size(); i++) {
-        if (outputMap[i].get_int() == n) {
-            return i;
-        }
-    }
-
-    throw std::logic_error("n is not present in outputmap");
-}
-
 AsyncRPCOperation_shieldcoinbase::AsyncRPCOperation_shieldcoinbase(
-        TransactionBuilder builder,
-        CMutableTransaction contextualTx,
-        std::vector<ShieldCoinbaseUTXO> inputs,
+        WalletTxBuilder builder,
+        ZTXOSelector ztxoSelector,
         PaymentAddress toAddress,
+        TransactionStrategy strategy,
+        int nUTXOLimit,
         CAmount fee,
         UniValue contextInfo) :
-        builder_(std::move(builder)), tx_(contextualTx), inputs_(inputs), fee_(fee), contextinfo_(contextInfo)
+    builder_(std::move(builder)),
+    ztxoSelector_(ztxoSelector),
+    toAddress_(toAddress),
+    strategy_(strategy),
+    nUTXOLimit_(nUTXOLimit),
+    fee_(fee),
+    contextinfo_(contextInfo)
 {
-    assert(contextualTx.nVersion >= 2);  // transaction format version must support vJoinSplit
-
-    if (fee < 0 || fee > MAX_MONEY) {
-        throw JSONRPCError(RPC_INVALID_PARAMETER, "Fee is out of range");
-    }
-
-    if (inputs.size() == 0) {
-        throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS, "Empty inputs");
-    }
-
-    //  Check the destination address is valid for this network i.e. not testnet being used on mainnet
-    examine(toAddress, match {
-        [&](CKeyID addr) {
-            throw JSONRPCError(RPC_VERIFY_REJECTED, "Cannot shield coinbase output to a p2pkh address.");
-        },
-        [&](CScriptID addr) {
-            throw JSONRPCError(RPC_VERIFY_REJECTED, "Cannot shield coinbase output to a p2sh address.");
-        },
```

Comment:
This change is about having shieldcoinbase only explicitly handle the behavior that is specific to it. The “can’t send to Sprout” policy is more general, and is handled in one place for all tx. This is only to check that shielding specifically doesn’t allow sending to transparent addresses. Although it could be clearer if it were encapsulated, like
```c++
bool RecipientIsTransparent(const PaymentAddress& address)
{
    return examine(address, …);
}

…

if (RecipientIsTransparent(destaddress.value())) {
    throw JSONRPCError(RPC_VERIFY_REJECTED, "Cannot shield coinbase output to a transparent address.");
}
```

Reply:
Oh, sorry, I misread this comment. I thought this was talking about the change _in_ rpcwallet. I agree it’s better for it to live here instead of there.